src/acl.c

```diff
@@ -1909,42 +1905,46 @@ int ACLCheckAllPerm(client *c, int *idxptr) {
     return ACLCheckAllUserCommandPerm(c->user, c->cmd, c->argv, c->argc, idxptr);
 }
 
-/* Check if the user's existing pub/sub clients violate the ACL pub/sub
- * permissions specified via the upcoming argument, and kill them if so. */
-void ACLKillPubsubClientsIfNeeded(user *new, user *original) {
+/* Check if the client should be killed because it is subscribed to channels that were
+ * permitted for the user `original` but not `new`. If `client` is NULL, this function
+ * will return 0 if-and-only-if _all_ channels permitted for `original` user are valid
+ * for `new` user. */
+int ACLShouldKillPubsubClient(user *new, user *original, client *c) {
     /* Do nothing if there are no subscribers. */
     if (!dictSize(server.pubsub_patterns) &&
         !dictSize(server.pubsub_channels) &&
         !dictSize(server.pubsubshard_channels))
-        return;
+        return 0;
 
     listIter li, lpi;
     listNode *ln, *lpn;
     robj *o;
-    int kill = 0;
-    
+
     /* First optimization is we check if any selector has all channel
      * permissions. */
     listRewind(new->selectors,&li);
     while((ln = listNext(&li))) {
         aclSelector *s = (aclSelector *) listNodeValue(ln);
-        if (s->flags & SELECTOR_FLAG_ALLCHANNELS) return;
+        if (s->flags & SELECTOR_FLAG_ALLCHANNELS) return 0;
     }
 
     /* Second optimization is to check if the new list of channels
      * is a strict superset of the original. This is done by
      * created an "upcoming" list of all channels that are in
      * the new user and checking each of the existing channels
      * against it.  */
-    list *upcoming = listCreate();
-    listRewind(new->selectors,&li);
-    while((ln = listNext(&li))) {
-        aclSelector *s = (aclSelector *) listNodeValue(ln);
-        listRewind(s->channels, &lpi);
-        while((lpn = listNext(&lpi))) {
-            listAddNodeTail(upcoming, listNodeValue(lpn));
+    if (!new->channels_cache) {
```

Comment:
I think it's an anti-pattern to have a cached value that isn't really intended to be viable long term. I sort of get why it's here, but I think it would be conceptually simpler to have a slightly different design. 

Instead, I think we should implement a function like:
```
list *getUpcomingChannelList(user *original, user* new) {
    if new can access all channels original return NULL;
    else return a list of channels that the new user can access
}
```
Which implements most of the logic of your current ACLShouldKillPubsubClient, however it will return the list of upcoming channels instead of caching it on the user. Next update ACLShouldKillPubsubClient so that it takes in a list of channels and returns 1 if the client should be killed based off that updated list.
```
int ACLShouldKillPubsubClient(client *c, list *channels) {
    if c is accessing any of the channels, return 1, else 0
}
```

In ACLLoadFromFile, we can maintain a RAX of user -> channel list, so that we can quickly re-look up the value. This is the re-implementation of your cache, but in a way that is local to the function. We expect the number of users to be small, so the channel lookup should still be comparably as fast.

Ideally this means that there is no user->cached_channel, so we never have to maintain it or clean it up.

Reply:
@madolson @slavak i made the above suggested changes. indeed seems better, please review.